### DISCUSSION TITLE
null-preventing operator

### DISCUSSION THREAD

**Discussion by VardanKosakyan** [dotnet/csharplang#5495] [Language Ideas]
**null-preventing operator (!?)**

Returns left operand if it is null otherwise it evaluates right operand and returns its result.
The proposed syntax is.

var r = a !? b;

Returns null if a is null, otherwise returns b. 
Like with ‘??’ Multiple operators can be combined.

var r = a !? b !? c;

By its functionality it is opposite to null-coalescing operator.
However, it can be useful in situations when right side expression shouldn`t be evaluated at all in case the left side operand is null.

Var r = nullableObject !? DoSomething( nullableObject);

**Motivation** 

Evaluating some expressions based on whether other objects are null or not is a common scenario.
Normally we do it this way

if(customer != null && order  != null)
{
    DoSomething(customer.Name, order.Number);
}

When we wish to assign a result, we may do it the following way:

var result = customer != null && order != null ? $”customer:{customer.Name}, order: {order.Number}” : null;

Proposed null-preventing operator allows doing the same using the following syntax: 

var result = customer !? order !? $”customer:{customer.Name}, order:{order.Number}”;

The expression with the traditional null-propagation operator is shorter, however, it does not give the desired results, when the target expression should not be evaluated at all (the expression is complex or it contains a method call). It is useful in simple scenarios.

var result = customer?.Name;

On the other hand, the proposed null-preventing operator is logically opposite to the null-coalescing operator (??). From a conceptual point of view, it is easily perceived.

**Alternatives**
//null checking operator
var result = customer != null && order != null ? $”customer:{customer.Name}, order:{order.Number}” : null;

//using helper method
var result2 = !Helper.IsAnyNull(customer,  order) ? $”customer:{customer.Name}, order:{order.Number}” : null;

**Comment by huoyaoyuan** [comment]
It behaves more like `?.` with extension method.

**Reply by RikkiGibson** [reply]
I am pretty sure we have seen proposals for operators like this in the past. In some ways it's just a generalization of `?.`: currently it is easy to "access a member on a receiver if it is not null", but it is currently still relatively clunky to generally "do something with a certain input if it is not null".

The examples that just look like `a !? b` do not look compelling by themselves for this reason 😄 the interesting scenarios are the ones like `x !? UseNotNullResult(x)`.

The trouble is that once the left side is more complex than a local variable read or something like that, it lends itself to repeating work or verbosity that is undesirable. `x.y.z() !? UseNotNullResult(x.y.z())` seems undesirable for example. Perhaps it would make sense to have a form that can add a temp containing the result of the left side: `x.y.z() as var w !? UseNotNullResult(w)` (strawman syntax). However, when you compare this "complex" case `x.y.z() as var w !? UseNotNullResult(w)` to the currently possible form `x.y.z() is { } w ? UseNotNullResult(w) : null`, it doesn't end up feeling like an obvious win.

**Reply by RikkiGibson** [reply]
edit: I was wrong in this post :), and don't currently have any suggestions or preferences on what the syntax for this operator would be.

~~If we were to consider an operator like this, I would probably prefer `?!` over `!?` for two reasons:~~
~~1. It more closely resembles its `?.` cousin~~
~~2. It doesn't introduce as much parsing difficulty: although `x! ? y : z` is weird, it currently exists and we need to know how to parse it. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4FgBQABALAAgFkAKYAezIBt9gBKLAby3xf2wEYA6AYTJgGcqAU04B1AE5QALkNIBCfAH58Acggr8IVcBW0A3FgC+WIA=). If we have `x ?! M(x)` I am pretty sure it will be simpler. There would be no need to scan ahead looking for a colon or something like that.~~

**Reply by siegfriedpammer** [reply]
I am not so sure about 2. see https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4FgBQABALAAgFkAKAe2ACsBTAYwBd9gBKLAby30/2wEYA6AMKkYAZ1IAbKnwDqAJyh0qxYAEIAvDACu48fgD8+FQCFlTfCHwAzCOJFUmAbiwBfLFmCkJ+E+Wr18AB5magB8+HSymlQOQA===

**Reply by RikkiGibson** [reply]
Oops, thanks. Yes, a very similar issue arises. :)

**Reply by VardanKosakyan** [reply]
proposed operator is logically opposite to the null-coalescing operator, therefore the "!?" syntax was suggested, since "!" character means  "inversion".  I agree   
x.y.z() is { } w ? UseNotNullResult(w) : null
syntax is still complex, especially in case of nested ternary.
